server/auth/auth.go

```diff
@@ -24,18 +24,270 @@ import (
+		return err
+	}
+
+	userPath := path.Join(userPrefix, name)
+
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	// Delete user from kv.
+	err = m.kv.Remove(userPath)
+	if err != nil {
+		return err
+	}
+
+	// Delete user from memory cache.
+	delete(m.users, name)
+
+	return nil
+}
+
+// ChangePassword changes password of a user.
+func (m *userManager) ChangePassword(name string, password string) error {
+	user, err := m.GetUser(name)
+	if err != nil {
+		return err
+	}
+
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	user.Hash = GenerateHash(password)
+
+	// Update user in kv
+	userJSON, err := json.Marshal(user)
+	if err != nil {
+		return errs.ErrJSONMarshal.Wrap(err).GenWithStackByCause()
+	}
+	userPath := path.Join(userPrefix, name)
+
+	err = m.kv.Save(userPath, string(userJSON))
+	if err != nil {
+		return err
+	}
+
+	// No need to update user in memory cache again.
+	return nil
+}
+
+// SetRoles sets roles of a user.
+func (m *userManager) SetRoles(name string, roles []string) error {
+	user, err := m.GetUser(name)
+	if err != nil {
+		return err
+	}
+
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	user.RoleKeys = roles
+
+	// Update user in kv
+	userJSON, err := json.Marshal(user)
+	if err != nil {
+		return errs.ErrJSONMarshal.Wrap(err).GenWithStackByCause()
+	}
+	userPath := path.Join(userPrefix, name)
+
+	err = m.kv.Save(userPath, string(userJSON))
+	if err != nil {
+		return err
+	}
+
+	// No need to update user in memory cache again.
+	return nil
+}
+
+// AddRole adds a role to a user.
+func (m *userManager) AddRole(name string, role string) error {
+	user, err := m.GetUser(name)
+	if err != nil {
+		return err
+	}
+
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	if ok := user.appendRole(role); !ok {
+		return errs.ErrUserHasRole.FastGenByArgs(name, role)
+	}
+
+	userJSON, err := json.Marshal(role)
+	if err != nil {
+		return errs.ErrJSONMarshal.Wrap(err).GenWithStackByCause()
+	}
+	userPath := path.Join(userPrefix, name)
+
+	// Update user in kv.
+	err = m.kv.Save(userPath, string(userJSON))
+	if err != nil {
+		return err
+	}
+
+	// Update user in memory cache.
+	m.users[name] = user
+
+	return nil
+}
+
+// RemoveRole removes a role from a user.
+func (m *userManager) RemoveRole(name string, role string) error {
+	user, err := m.GetUser(name)
+	if err != nil {
+		return err
+	}
+
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	if ok := user.removeRole(role); !ok {
+		return errs.ErrUserMissingRole.FastGenByArgs(name, role)
+	}
+
+	userJSON, err := json.Marshal(role)
+	if err != nil {
+		return errs.ErrJSONMarshal.Wrap(err).GenWithStackByCause()
+	}
+	userPath := path.Join(userPrefix, name)
+
+	// Update user in kv.
+	err = m.kv.Save(userPath, string(userJSON))
+	if err != nil {
+		return err
+	}
+
+	// Update user in memory cache.
+	m.users[name] = user
+
+	return nil
+}
+
+// UpdateCache refreshes in-memory cache of users.
+func (m *userManager) UpdateCache() error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	userPath := strings.Join([]string{userPrefix, ""}, "/")
```

Comment:
I see different methods to build userPath

Reply:
Using this method in `GetUserPath` because it can handle trailing slash correctly.